### PR TITLE
Add new EM_IMPORT macro that adds `import_name` and `import_module` attributes

### PR DESCRIPTION
**Pull request by sbc100**
This allows us to explicitly mark function that are imported from the
outside world (i.e. JS).  Hopefully one day this should allow us to
remove the `--allow-undefined` linker flag.
    
This change is enough to allow hello world to build without
`--allow-undefined`

### CODE REVIEW

**Comment by sbc100** [comment]
I wanted to call this macro EM_IMPORT but I know jacob has plans to use that name.  Thoughts?

**Comment by kripken** [comment]
Can you please explain more what the idea is here? I don't know much about the linker side, and why adding "env" (since the name is already known?) helps avoid needing that linker flag.

If this just adds the module name, perhaps we can tell the linker that imports are from "env" once as a flag, instead of tagging all the headers many times? (Probably not a good idea for some reason...)

**Comment by sbc100** [comment]
I could drop the "env" since the linker already treats it as a default.     I'd like the linker to so stop using that default eventually and indeed we could have a flag to specify the default.   However specifying per symbol doesn't have any downside if its embedded in a macro does it?   I do think the macro needs a better name though.     How about EM_JS_IMPORT?

**Comment by sbc100** [comment]
The idea here is to pre-declare all our JS functions in this way one day, so that wasm-ld can be away all symbols out there.   The goal would be to one day drop the `--allow-undefined` flag and have the linker report nice error messages.

File: system/include/emscripten/emscripten.h [hunk 1]
```diff
@@ -60,6 +60,13 @@ typedef void (*em_str_callback_func)(const char *);
 
 #define EMSCRIPTEN_KEEPALIVE __attribute__((used))
 
+#ifdef __wasm__
+#define EM_ENV_IMPORT(NAME) __attribute__((import_module("env"))) \
```
Comment (jgravelle-google) [inline system/include/emscripten/emscripten.h]: Nit: can do `__attr__((a, b))` instead of `__attr__((a)) __attr((b))`

**Comment by kripken** [comment]
I see, thanks @sbc100 ! sgtm (and no feelings about the name)

**Comment by sbc100** [comment]
Since I'm imaging that eventually all users who want to import JS library function will need to use this mechanism I'd rather not use the longer `EMSCRIPTEN_JS_SYSTEM_IMPORT`.

Perhaps EM_JS_LIBRARY_IMPORT?   I still prefer EM_IMPORT.. so I might still that name for now.  Since this is just an experiment we can rename it later after discussion.

**Comment by sbc100** [comment]
I could do EM_IMPORT_AS(...) maybe?

File: system/include/emscripten/emscripten.h (hunk 1 above)
Comment (sbc100) [reply system/include/emscripten/emscripten.h]: Done.

**Comment by sbc100** [comment]
How about EM_EXTERN?

I'm not sure the name matters too much right now, at least not until external users start using the macro.

File: ChangeLog.md [hunk 2]
```diff
@@ -22,6 +22,10 @@ Current Trunk
   However `emcc`, `emcc.py` and `python emcc.py` all continue to work.
   The reason for this change is that `#!/usr/bin/env python` is no longer
   portable since the python symlink was dropped from Ubuntu 20.04.
+- New EM_IMPORT macro to make C/C++ symbols are imported from outside the module
+  (i.e. imported from JS).  Currently we still assume that all undefined 
+  symbols can come from JS, but in the future we hope to mark all such symbols
+  in order to allow wasm-ld to report actual undefined symbols.
```
Comment (kripken) [inline ChangeLog.md]: Not sure I follow this text. Could the first sentence be `New EM_IMPORT macro that sets the module of a wasm import for an external C function` - ? And in the second, should `can come` be `must come`? And `all such` refers to what (things from JS, or all externals, or something else)?

**Comment by sbc100** [comment]
I re-phrases the changelog.  Does it make more sense now?

File: ChangeLog.md [hunk 3]
```diff
@@ -22,6 +22,11 @@ Current Trunk
   However `emcc`, `emcc.py` and `python emcc.py` all continue to work.
   The reason for this change is that `#!/usr/bin/env python` is no longer
   portable since the python symlink was dropped from Ubuntu 20.04.
+- New EM_IMPORT macro to mark C/C++ symbols as imported from outside the module
+  (i.e. imported from JS).  Currently we still default to assuming that *all*
```
Comment (kripken) [inline ChangeLog.md]: Should that `i.e.` be an `e.g.` as in dynamic linking it might be another wasm module?

Comment (kripken) [reply ChangeLog.md]: but overall this seems clear enough to me, lgtm.

**Comment by Qix-** [comment]
Sorry for the gravedig; using `import_name` and `import_module` still result in undefined symbols. Is there something that needs to be passed to `emcc` to pick them up as imports and not undefined symbols?

**Comment by sbc100** [comment]
Oh I see, this is not wasm-ld issue, this is emscripten noticing that `panic` is being (correctly) imported by the wasm module but emscripten's JavaScript library doesn't contain a `panic` symbol to provide to the module so its reporting its undefined in the entire program.

The way emscripten works is that JavaScript library code must, by default, provide definitions for anything that is imported by the wasm module.    Otherwise there is no way the JavaScript glue code can successfully load the wasm module.

Now, should be behave differently in STANDALONE_WASM mode where the JavaScript code it not relevant?   Maybe.. I can't remember right now why we do any of the JS library processing at all in this mode.

**Comment by Qix-** [comment]
I figured it might be something like that - thanks for the info. :)

Yeah I'm not using any of Emscripten's javascript stuff. I'd rather just use the out-of-the-box clang WASM tooling (as it works fine) but emscripten has support for all of the exceptions, allocation, etc. stuff which I need for some libraries. My WASM binary will be executed in multiple environments, not just Javascript.

Is this something worth opening a new issue over?

**Comment by sbc100** [comment]
Yes, perhaps open a new issue.

**Pull request merged**